### PR TITLE
Python API usability issues.  Changes:

### PR DESCRIPTION
**Pull request by luketpeterson**
1. Changed names to be more clear how different objects should be used. 
 - PySpace -> Abstract
 - Space Space -> SpaceRef
 - GroundingSpace -> GroundingSpaceRef

2. Added API comments, briefly describing intended usage of objects and functions

3. Changed Space `query` return value to BindingsSet in both C and Python APIs

4. Added BindingsSet implementation for `__getitem__ ` Python BindingsSet, in order to be able to access the individual Bindings in the same list-of-dict format formerly returned by space.query()

5. Added `GroundingSpace` Python wrapper implementation, so basic GroundingSpace can be easily subclassed.

6. Rename hyperonpy trampoline functions to `_priv_` so users don’t get confused

### CODE REVIEW

**Comment by luketpeterson** [comment]
Hopefully this addresses the issues raised in https://github.com/trueagi-io/hyperon-experimental/issues/335

**Review by vsbogd — APPROVED** [review]
Thanks Luke. To me this PR more clearly distinguishes `AbstractSpace <- SpecificSpace` and `SpaceRef <- GroundingSpaceRef` hierarchies, thus I approve it. I think we should also wait for @Necr0x0Der's opinion before merge.

**Comment by Necr0x0Der** [comment]
I agree that this is an improvement. But my main request was to introduce a class that can reuse the existing space and redefine only `query`. Does the current PR suppose that one can inherit from `GroundingSpace` and do `kb = SpaceRef(MyGroundingSpace)`, where `MyGroundingSpace` only redefines `query`? If no, what is a supposed way to do this?

File: python/hyperon/__init__.py
Comment (luketpeterson) [reply python/hyperon/__init__.py]: That is much cleaner.  Changed in 8951284

**Comment by luketpeterson** [comment]
> I agree that this is an improvement. But my main request was to introduce a class that can reuse the existing space and redefine only `query`. Does the current PR suppose that one can inherit from `GroundingSpace` and do `kb = SpaceRef(MyGroundingSpace)`, where `MyGroundingSpace` only redefines `query`? If no, what is a supposed way to do this?

I added the `test_python_extended_grounding_space` test to the `test_grounding_space.py` file in the d04983c commit.  Is this what you have in mind?  Or did I misunderstand?

**Comment by Necr0x0Der** [comment]
> I added the `test_python_extended_grounding_space` test to the `test_grounding_space.py` file in the [d04983c](https://github.com/trueagi-io/hyperon-experimental/commit/d04983cf6decaf4bb05ced6046694b845135d9c9) commit. Is this what you have in mind? Or did I misunderstand?

Yes, exactly. This is what I wanted to confirm.

**Review by Necr0x0Der — APPROVED** [review]

**Pull request merged**
